src/L1/rollup/ScrollChain.sol

```diff
@@ -569,24 +621,121 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
+        bytes32 cachedStateRoot = finalizedStateRoots[_batchIndex];
+        if (cachedStateRoot == bytes32(0) || _batchIndex > lastZkpVerifiedBatchIndex) {
+            revert ErrorBatchNotVerifiedWithZKProof();
+        }
+
+        uint256 _lastVerifiedBatchIndex = lastTeeVerifiedBatchIndex;
+        if (_batchIndex <= _lastVerifiedBatchIndex) {
+            revert ErrorBatchIsAlreadyVerified();
+        }
+
+        bytes memory _publicInput = abi.encodePacked(
+            layer2ChainId,
+            uint32(_batchIndex - _lastVerifiedBatchIndex), // numBatches
+            finalizedStateRoots[_lastVerifiedBatchIndex], // _prevStateRoot
+            committedBatches[_lastVerifiedBatchIndex], // _prevBatchHash
+            _postStateRoot,
+            _batchHash,
+            _withdrawRoot
+        );
+
+        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
+        IRollupVerifier(teeVerifier).verifyBundleProof(batchVersion, _batchIndex, _teeProof, _publicInput);
+
+        // check withdraw root and state root match with zk proof
+        if (withdrawRoots[_batchIndex] != _withdrawRoot || cachedStateRoot != _postStateRoot) {
+            unresolvedState.batchIndex = _batchIndex;
+            unresolvedState.stateRoot = _postStateRoot;
+            unresolvedState.withdrawRoot = _withdrawRoot;
+            emit StateMismatch(_batchIndex, _postStateRoot, _withdrawRoot);
+            return;
+        }
+
+        lastTeeVerifiedBatchIndex = _batchIndex;
+
         // Pop finalized and non-skipped message from L1MessageQueue.
         _finalizePoppedL1Messages(_totalL1MessagesPoppedOverall);
 
+        emit VerifyBatchWithTee(_batchIndex);
         emit FinalizeBatch(_batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
     }
 
     /************************
      * Restricted Functions *
      ************************/
 
+    /// @notice Resolve mismatched state.
+    ///
+    /// @dev This should only be called by Security Council.
+    ///
+    /// @param useSGXState Whether we want to use the state root from SGX prover.
+    function resolveStateMismatch(bytes calldata _batchHeader, bool useSGXState) external onlyOwner {
+        UnresolvedState memory state = unresolvedState;
+        if (state.batchIndex == 0) {
+            revert ErrorNoUnresolvedState();
+        }
+        (, , uint256 _batchIndex, uint256 _totalL1MessagesPoppedOverall) = _loadBatchHeader(_batchHeader);
+        if (_batchIndex != state.batchIndex) revert ErrorBatchIndexMismatch();
+        if (useSGXState) {
+            finalizedStateRoots[state.batchIndex] = state.stateRoot;
+            withdrawRoots[state.batchIndex] = state.withdrawRoot;
+            emit ResolveState(state.batchIndex, state.stateRoot, state.withdrawRoot);
+
+            // reset zkp verified batch index, zk prover need to reprove everything after this batch
+            lastZkpVerifiedBatchIndex = state.batchIndex;
```

Comment:
@colinlyguo any thoughts what we'd need to add to be able to recover from this relatively painlessly? Maybe just listen to `ResolveState` events in `rollup-relayer` and update db accordingly? (Though in this case most likely we'll need to fix either zk or sgx prover.)

Reply:
> Maybe just listen to ResolveState events in rollup-relayer and update db accordingly